### PR TITLE
EVG-8219, EVG-8222, EVG-8223: Implemented DiskUsage, Uptime, and Process collectors

### PR DESCRIPTION
**Pull request by maravichandran**

### CODE REVIEW

File: agent/system_metrics.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	VSZ           uint64         `json:"vsz"`
+	RSS           uint64         `json:"rss"`
+	TT            string         `json:"tt"`
+	Stat          string         `json:"stat"`
+	Started       int64          `json:"started"`
+	Time          *cpu.TimesStat `json:"time"`
+	Command       string         `json:"command"`
+}
+
+func (collector *DiskUsageCollector) Name() string {
```
Comment (julianedwards) [inline agent/system_metrics.go:86]: a style nit: you can make these functions one-liners since they so small

File: agent/system_metrics.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	connOpts    system_metrics.ConnectionOptions
+	interval    time.Duration
+	collectors  []MetricCollector
+	id          string
+	client      *system_metrics.SystemMetricsClient
+	errored     bool
+	closed      bool
+}
+
+type DiskUsageCollector struct{}
```
Comment (julianedwards) [inline agent/system_metrics.go:47]: Is there a reason these three structs are public?

File: agent/system_metrics.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	DataFormatBSON DataFormat = 2
+	DataFormatJSON DataFormat = 3
+	DataFormatCSV  DataFormat = 4
+)
+
+type UptimeWrapper struct {
+	Uptime uint64 `json:"uptime"`
+}
+
+type ProcessesWrapper struct {
```
Comment (julianedwards) [inline agent/system_metrics.go:69]: This can't have the same name as the following struct

File: agent/system_metrics.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+			TT:            terminal,
+			Stat:          status,
+			Started:       createTime,
+			Time:          times,
+			Command:       name,
+		}
+		processesPopulated[i] = processWrapper
+	}
+	processesWrapper := ProcessesWrapper{processesPopulated}
+	return json.Marshal(processesWrapper)
```
Comment (julianedwards) [inline agent/system_metrics.go:188]: wrap the error with a message

File: agent/system_metrics.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+func (collector *ProcessCollector) Collect(ctx context.Context) ([]byte, error) {
+	processes, err := process.Processes()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	processesPopulated := make([]ProcessWrapper, len(processes))
+
+	for i, process := range processes {
+		cpuPercent, err := process.CPUPercent()
```
Comment (julianedwards) [inline agent/system_metrics.go:138]: should we be tracking these errors?

Comment (maravichandran) [reply agent/system_metrics.go:138]: A lot of them fail, I think for reasons beyond our control (e.g. the process ended in the middle of requesting the information). I don't think we need to track the errors

File: agent/system_metrics.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	uptimeWrapper := UptimeWrapper{uptime}
+	return convertJSONToFTDC(ctx, uptimeWrapper)
+}
+
+func (collector *ProcessCollector) Collect(ctx context.Context) ([]byte, error) {
+	processes, err := process.Processes()
```
Comment (Kimchelly) [inline agent/system_metrics.go:130]: nit: can we shorten this to "procs"? Go prefers shorter names as long as they don't interfere with understanding the code.

File: agent/system_metrics.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	processesPopulated := make([]ProcessWrapper, len(processes))
```
Comment (Kimchelly) [inline agent/system_metrics.go:135]: Can we call this `procMetrics`?

File: agent/system_metrics.go (hunk 1 above)
Comment (maravichandran) [reply agent/system_metrics.go:86]: Made them one line

File: agent/system_metrics.go (hunk 3 above)
Comment (maravichandran) [reply agent/system_metrics.go:69]: done

File: agent/system_metrics.go (hunk 2 above)
Comment (maravichandran) [reply agent/system_metrics.go:47]: done

File: agent/system_metrics.go (hunk 4 above)
Comment (maravichandran) [reply agent/system_metrics.go:188]: done

File: glide.yaml [hunk 8]
```diff
@@ -57,7 +57,6 @@ import:
   - package: github.com/evergreen-ci/pail
   - package: gonum.org/v1/gonum
   - package: github.com/evergreen-ci/timber
-
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply glide.yaml:60]: done

File: agent/system_metrics.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+type ProcessWrapper struct {
+	Pid           int32          `json:"pid"`
+	CPUPercent    float64        `json:"%cpu"`
+	MemoryPercent float32        `json:"%mem"`
+	VSZ           uint64         `json:"vsz"`
+	RSS           uint64         `json:"rss"`
+	TT            string         `json:"tt"`
+	Stat          string         `json:"stat"`
+	Started       int64          `json:"started"`
+	Time          *cpu.TimesStat `json:"time"`
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go:82]: changed it to not a pointer

File: agent/system_metrics.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+	DataFormatJSON DataFormat = 3
+	DataFormatCSV  DataFormat = 4
+)
+
+type UptimeWrapper struct {
+	Uptime uint64 `json:"uptime"`
+}
+
+type ProcessesWrapper struct {
+	Processes []ProcessWrapper
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go:70]: done

File: agent/system_metrics.go (hunk 6 above)
Comment (maravichandran) [reply agent/system_metrics.go:130]: done

File: agent/system_metrics.go (hunk 7 above)
Comment (maravichandran) [reply agent/system_metrics.go:135]: done

File: agent/system_metrics.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+func (collector *DiskUsageCollector) Name() string {
+	return "disk_usage"
+}
+
+func (collector *UptimeCollector) Name() string {
+	return "uptime"
+}
+
+func (collector *ProcessCollector) Name() string {
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go]: done

File: agent/system_metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go]: Something like this? 

```
func NewDiskUsageCollector() *diskUsageCollector {
	return new(diskUsageCollector)
}
```

File: agent/system_metrics.go [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+type UptimeWrapper struct {
+	Uptime uint64 `json:"uptime"`
+}
+
+type ProcessesWrapper struct {
+	Processes []ProcessWrapper
+}
+
+type ProcessWrapper struct {
+	Pid           int32          `json:"pid"`
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go]: done

File: agent/system_metrics_test.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+	assert := assert.New(t)
+	ctx := context.TODO()
+	uptimeCollector := &UptimeCollector{}
+	output, err := uptimeCollector.Collect(ctx)
+	assert.NoError(err)
+
+	iter := ftdc.ReadMetrics(ctx, bytes.NewReader(output))
+	i := 0
+	for iter.Next() {
+		i++
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics_test.go:49]: done

File: agent/system_metrics_test.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+	assert := assert.New(t)
+	ctx := context.TODO()
+	diskUsageCollector := &DiskUsageCollector{}
+	output, err := diskUsageCollector.Collect(ctx)
+	assert.NoError(err)
+
+	iter := ftdc.ReadMetrics(ctx, bytes.NewReader(output))
+	i := 0
+	for iter.Next() {
+		i++
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics_test.go:23]: done

File: agent/system_metrics.go [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+func (collector *UptimeCollector) Format() DataFormat {
+	return DataFormatFTDC
+}
+
+func (collector *ProcessCollector) Format() DataFormat {
+	return DataFormatJSON
+}
+
+func (collector *DiskUsageCollector) Collect(ctx context.Context) ([]byte, error) {
+	usage, err := disk.Usage("/")
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go]: done

File: agent/system_metrics.go [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+	Processes []ProcessWrapper
+}
+
+type ProcessWrapper struct {
+	Pid           int32          `json:"pid"`
+	CPUPercent    float64        `json:"%cpu"`
+	MemoryPercent float32        `json:"%mem"`
+	VSZ           uint64         `json:"vsz"`
+	RSS           uint64         `json:"rss"`
+	TT            string         `json:"tt"`
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go]: done

File: agent/system_metrics.go [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+
+func convertJSONToFTDC(ctx context.Context, metric interface{}) ([]byte, error) {
+	jsonMetrics, err := json.Marshal(metric)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem converting metrics to JSON")
+	}
+
+	opts := metrics.CollectJSONOptions{
+		InputSource:   bytes.NewReader(jsonMetrics),
+		SampleCount:   100,
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go:193]: changed to 1

File: agent/system_metrics.go [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+		FlushInterval: 100 * time.Millisecond,
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go:194]: Changed to 1 second

File: agent/system_metrics.go [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+	processes, err := process.Processes()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	processesPopulated := make([]ProcessWrapper, len(processes))
+
+	for i, process := range processes {
+		cpuPercent, err := process.CPUPercent()
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics.go:131]: done

File: agent/system_metrics_test.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+	}
+	assert.Equal(1, i)
+}
+
+func TestCollectProcesses(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply agent/system_metrics_test.go:55]: The output is a marshaled JSON rather than FTDC

File: glide.yaml [hunk 21]
```diff
@@ -57,8 +57,11 @@ import:
...
   - package: github.com/evergreen-ci/certdepot
+  - package: github.com/shirou/gopsutil
+  - package: github.com/StackExchange/wmi
+  - package: github.com/go-ole/go-ole
+  - package: github.com/shirou/w32
```
Comment (Kimchelly) [inline glide.yaml:64]: Is there a reason we need these other packages? Are they dependencies of gopsutil?

File: agent/system_metrics.go [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	uptimeWrapper := UptimeWrapper{uptime}
+	return convertJSONToFTDC(ctx, uptimeWrapper)
+}
+
+type processCollector struct{}
+
+type ProcessesWrapper struct {
```
Comment (Kimchelly) [inline agent/system_metrics.go:120]: Same comment about not exporting.

File: agent/system_metrics.go [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+	output, err := metrics.CollectJSONStream(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem converting FTDC to JSON")
+	}
+	return output, nil
+}
+
+type uptimeCollector struct{}
+
+type UptimeWrapper struct {
```
Comment (Kimchelly) [inline agent/system_metrics.go:100]: Unless we have a reason to export this, we shouldn't export it since it's not used outside this package.

File: agent/system_metrics.go [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+	DataFormatText DataFormat = 0
+	DataFormatFTDC DataFormat = 1
+	DataFormatBSON DataFormat = 2
+	DataFormatJSON DataFormat = 3
+	DataFormatCSV  DataFormat = 4
+)
+
+type diskUsageCollector struct{}
+
+func NewDiskUsageCollector() *diskUsageCollector {
```
Comment (Kimchelly) [inline agent/system_metrics.go:62]: Add comment.

File: glide.yaml (hunk 21 above)
Comment (maravichandran) [reply glide.yaml:64]: For some reason the Windows tests threw an error because these dependencies were missing so I added them

File: agent/system_metrics.go (hunk 22 above)
Comment (maravichandran) [reply agent/system_metrics.go:120]: done

File: agent/system_metrics.go (hunk 23 above)
Comment (maravichandran) [reply agent/system_metrics.go:100]: done

File: agent/system_metrics.go (hunk 24 above)
Comment (maravichandran) [reply agent/system_metrics.go:63]: done

File: agent/system_metrics.go [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+// NewProcessCollector creates a processCollector object.
+func NewProcessCollector() *processCollector {
+	return new(processCollector)
+}
+
+type processesWrapper struct {
+	Processes []ProcessData `json:"processes"`
+}
+
+type ProcessData struct {
```
Comment (julianedwards) [inline agent/system_metrics.go:135]: Do we need to export this? If so, add a comment.

File: glide.yaml (hunk 21 above)
Comment (Kimchelly) [reply glide.yaml:64]: Can you add a comment above gopsutil saying `# gopsutil and its dependencies`?

File: agent/system_metrics.go (hunk 25 above)
Comment (Kimchelly) [reply agent/system_metrics.go:135]: It shouldn't need to be exported.

File: agent/system_metrics.go [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+)
+
+type diskUsageCollector struct{}
+
+// NewDiskUsageCollector creates a diskUsageCollector object.
+func NewDiskUsageCollector() *diskUsageCollector {
+	return new(diskUsageCollector)
+}
+
+func (collector *diskUsageCollector) Name() string { return "disk_usage" }
```
Comment (Kimchelly) [inline agent/system_metrics.go:67]: nit: the method receiver should be a short name like `c` instead of `collector`. Usually method receivers are short (ideally a few characters long).

File: agent/system_metrics.go [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+func convertJSONToFTDC(ctx context.Context, metric interface{}) ([]byte, error) {
+	jsonMetrics, err := json.Marshal(metric)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem converting metrics to JSON")
+	}
+
+	opts := metrics.CollectJSONOptions{
+		InputSource:   bytes.NewReader(jsonMetrics),
+		SampleCount:   1,
+		FlushInterval: 1 * time.Second,
```
Comment (Kimchelly) [inline agent/system_metrics.go:89]: nit: This can just be `time.Second`.

File: agent/system_metrics.go [hunk 28]
```diff
@@ -0,0 +1,226 @@
...
+	return new(processCollector)
+}
+
+type processesWrapper struct {
+	Processes []ProcessData `json:"processes"`
+}
+
+type ProcessData struct {
+	PID               int32   `json:"pid"`
+	CPUPercent        float64 `json:"%cpu"`
```
Comment (Kimchelly) [inline agent/system_metrics.go:137]: Can we keep % characters out of JSON tags?

File: agent/system_metrics.go (hunk 25 above)
Comment (Kimchelly) [inline agent/system_metrics.go:135]: Can we add `omitempty` to the JSON tags so that they don't get included if they're not set?

File: glide.yaml (hunk 21 above)
Comment (maravichandran) [reply glide.yaml:64]: done

File: agent/system_metrics.go (hunk 25 above)
Comment (maravichandran) [reply agent/system_metrics.go:135]: changed to unexported

File: agent/system_metrics.go (hunk 26 above)
Comment (maravichandran) [reply agent/system_metrics.go:67]: done

File: agent/system_metrics.go (hunk 27 above)
Comment (maravichandran) [reply agent/system_metrics.go:89]: done

File: agent/system_metrics.go (hunk 25 above)
Comment (maravichandran) [reply agent/system_metrics.go:135]: done

File: agent/system_metrics.go (hunk 28 above)
Comment (maravichandran) [reply agent/system_metrics.go:137]: done

**Comment by Kimchelly** [comment]
This was merged in #3870.
